publication-commons/src/main/java/no/unit/nva/doi/models/ImmutableDoi.java

```diff
@@ -0,0 +1,369 @@
+    }
+
+    /**
+     * Copy the current immutable object by setting a value for the {@link Doi#getSuffix()} () suffix} attribute. An
+     * equals check used to prevent copying of the same value by returning {@code this}.
+     *
+     * @param value A new value for suffix
+     * @return A modified copy of the {@code this} object
+     */
+    public final ImmutableDoi withSuffix(String value) {
+        String newValue = Objects.requireNonNull(value, MESSAGE_NON_NULL_ARGUMENT_FOR_PARAMETER_SUFFIX);
+        if (this.suffix.equals(newValue)) {
+            return this;
+        }
+        return new ImmutableDoi(this.proxy, this.prefix, newValue);
+    }
+
+    /**
+     * This instance is equal to all instances of {@code ImmutableDoi} that have equal attribute values.
+     *
+     * @return {@code true} if {@code this} is equal to {@code another} instance
+     */
+    @Override
+    public boolean equals(Object another) {
+        if (this == another) {
+            return true;
+        }
+        return another instanceof ImmutableDoi
+            && equalTo((ImmutableDoi) another);
+    }
+
+    /**
+     * Computes a hash code from attributes: {@code proxy}, {@code prefix}, {@code suffix}.
+     *
+     * @return hashCode value
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hash(proxy, prefix, suffix);
+    }
+
+    /**
+     * Prints the immutable value {@code Doi} with attribute values.
+     *
+     * @return A string representation of the value
+     */
+    @Override
+    public String toString() {
+        return getPrefix() + FORWARD_SLASH + getSuffix();
+    }
+
+    private static void validateProxyUri(URI proxy) {
+        try {
+            proxy.toURL();
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException(ERROR_PROXY_URI_MUST_BE_A_VALID_URL, e);
+        }
+    }
+
+    private static URI createUriWithoutPathQueryAndFragmentWithSuffixForwardSlash(URI value) {
+        try {
+            return new URI(value.getScheme(), value.getUserInfo(), value.getHost(), value.getPort(),
+                FORWARD_SLASH_STRING,
+                null, null);
+        } catch (Exception e) {
+            throw new IllegalStateException("Could not reconstruct URI and strip path, query and fragment arguments");
+        }
+    }
+
+    private boolean equalTo(ImmutableDoi another) {
+        return proxy.equals(another.proxy)
+            && prefix.equals(another.prefix)
+            && suffix.equals(another.suffix);
+    }
+
+    /**
+     * Builds instances of type {@link ImmutableDoi ImmutableDoi}. Initialize attributes and then invoke the {@link
+     * #build()} method to create an immutable instance.
+     *
+     * <p><em>{@code Builder} is not thread-safe and generally should not be stored in a field or collection,
+     * but instead used immediately to create instances.</em>
+     */
+    public static final class Builder {
```

Comment:
This has sufficient complexity that I am lead to believe that there are simpler ways of achieving the same; that said, I am unsure what you are trying to achieve from the code.

Reply:
We replaced the bitwise optimization and does null checking instead. 

To move on and getting delivery of user story: https://github.com/BIBSYSDEV/nva-publication-api/pull/96#issuecomment-731022543